File: api/webhook.js
const twilio = require('twilio');
const { exec } = require('child_process');
const path = require('path');
const fs = require('fs');

const accountSid = process.env.TWILIO_ACCOUNT_SID;
const authToken = process.env.TWILIO_AUTH_TOKEN;
const client = twilio(accountSid, authToken);

const twilioWhatsAppNumber = 'whatsapp:+14155238886';

const ytDlpExecutable = '/opt/bin/yt-dlp';
const ffmpegLocation = '/opt/bin/ffmpeg';

module.exports = async (req, res) => {
  const incomingMessage = req.body.Body;
  const from = req.body.From;

  if (isYouTubeLink(incomingMessage)) {
    const videoUrl = incomingMessage;
    const outputFilename = 'audio.mp3';
    const outputPath = path.join(__dirname, outputFilename);

    try {
      if (fs.existsSync(outputPath)) {
        fs.unlinkSync(outputPath);
      }

      await downloadAudio(videoUrl, outputPath);

      const publicUrl = `https://${process.env.VERCEL_URL}/public/${outputFilename}`;

      await client.messages.create({
        from: twilioWhatsAppNumber,
        to: from,
        mediaUrl: [publicUrl],
      });

      res.status(200).send('Message sent');
    } catch (error) {
      console.error(error);
      await client.messages.create({
        from: twilioWhatsAppNumber,
        to: from,
        body: 'Sorry, there was an error processing your request.',
      });

      res.status(500).send('Error');
    }
  } else {
    await client.messages.create({
      from: twilioWhatsAppNumber,
      to: from,
      body: 'Please send a valid YouTube link.',
    });

    res.status(200).send('Invalid link');
  }
};

function isYouTubeLink(url) {
  return url.includes('youtube.com') || url.includes('youtu.be');
}

function downloadAudio(videoUrl, outputPath) {
  return new Promise((resolve, reject) => {
    const command = `"${ytDlpExecutable}" ${videoUrl} --extract-audio --audio-format mp3 --output "${outputPath}" --ffmpeg-location "${ffmpegLocation}"`;

    exec(command, (error, stdout, stderr) => {
      if (error) {
        console.error(`exec error: ${error}`);
        reject(error);
        return;
      }
      if (stderr) {
        console.error(`stderr: ${stderr}`);
        reject(stderr);
        return;
      }
      console.log(`stdout: ${stdout}`);
      resolve(stdout);
    });
  });
}

File: api/webhook.py
import yt_dlp
import os
from twilio.rest import Client

# Function to download audio from YouTube
def download_audio(video_url):
    ydl_opts = {
        'format': 'bestaudio/best',
        'extractaudio': True,  # download only audio
        'audioquality': 1,  # best quality
        'outtmpl': '/tmp/audio.mp3',  # output file path (Vercel allows /tmp)
    }

    with yt_dlp.YoutubeDL(ydl_opts) as ydl:
        ydl.download([video_url])
    
    return '/tmp/audio.mp3'

# Example Twilio integration (you should replace with your Twilio details)
def send_message(audio_path, to_phone_number):
    # Your Twilio credentials
    account_sid = 'your_account_sid'
    auth_token = 'your_auth_token'

    client = Client(account_sid, auth_token)

    # Send the audio file (you'll need to store it somewhere accessible or send it as an attachment)
    message = client.messages.create(
        body="Here is your audio file!",
        from_='+your_twilio_number',
        to=to_phone_number,
        media_url=f'http://example.com{audio_path}',  # You'll need to upload the file somewhere
    )

    return message.sid

# Sample webhook handler
def handler(request):
    video_url = request.json.get('video_url')
    phone_number = request.json.get('phone_number')

    # Download audio from the YouTube video
    audio_path = download_audio(video_url)

    # Send audio file via Twilio (ensure the file is accessible)
    send_message(audio_path, phone_number)

    return {"status": "success"}
